WIP: Added managed interactives [#172610374]

dougmartin opened the pull request.

dougmartin commented:
@scytacki I wasn't able to use a base interactive model class for both `MWInteractive` and `ManageInteractive` as the `constantize` method used when rendering pages returned the base class name which breaks the core page render.  I tried moving the base class code into a module but that caused other issues.  This is the cause of the duplicate code CodeClimate issues.

Now that I have a good test suite passing I'll look again tomorrow to see if I can refactor out the common code.

You'll also notice a few `TODO REVIEW` comments for questions I have.

scytacki commented on app/models/managed_interactive.rb:
```diff
@@ -0,0 +1,269 @@
+      :inherit_image_url,
+      :custom_image_url
+    ])
+    if library_interactive
```
It should always have a library interactive (at least based on code above) so this check could be removed.

dougmartin commented:
@scytacki I've added tests for import/export of managed interactives and library interactives using the export hash we discussed.  I was also able to use a common module after I removed one method from the module that called a method in the ManagedInteractive subclass and used an attribute with the same name in the MWInteractive.

scytacki commented on app/models/managed_interactive.rb:
```diff
@@ -0,0 +1,269 @@
+    managed_interactive
+  end
+
+  # REVIEW TODO: is update_labbook_options needed?
```
Lets take this out but put a comment in the MwInteractive code indicating this is not supported by the new ManagedInteractive. So we need to migrate this feature to a better solution.

I'll add a story about this in the hopes we can get to it before putting this down.

dougmartin replied:
Done.

dougmartin commented on app/models/managed_interactive.rb:
```diff
@@ -0,0 +1,269 @@
+  end
+
+  def enable_learner_state
+    library_interactive ? library_interactive.enable_learner_state : false
```
I've updated the model and spec to not assume the library interactive is set.  The user could add a managed interactive to the page but not select a library interactive.

scytacki commented on app/models/managed_interactive.rb:
```diff
@@ -0,0 +1,171 @@
+  end
+
+  # this can't be moved into BaseInteractive as enable_learner_state is defined as a method in
+  # this class and that causes reportable? not to be defined correctly for outside callers
```
Nevermind, I see this is a module not a class. So ignore this.

scytacki commented on app/models/managed_interactive.rb:
```diff
@@ -0,0 +1,171 @@
+
+        # REVIEW QUESTION: should we call #save here - if we don't and the user imports an activity
+        # that uses the same library interactive more than once it won't find the newly created
+        # library interactive
```
Go ahead with the save.  Just to know about it it would be good to look at the second pass code that is handling related linked interactives. And so we don't forget about this save when allow authors to import, add a comment to the top level import.

scytacki commented on app/controllers/managed_interactives_controller.rb:
```diff
@@ -0,0 +1,22 @@
+  end
+
+  # REVIEW: this was copied from the MWInteractivesController but it doesn't look to be used anywere
+  # def create_interactive
```
You can take it out.

dougmartin commented:
@scytacki I've addressed all the review comments.  I also double checked the managed interactives against the library interactive and mw interactive and found two unproxied attributes that I added.  I also added a test to cover the save of the library and note about the save in the import model.

scytacki commented on app/models/managed_interactive.rb:
```diff
@@ -33,59 +33,52 @@ class ManagedInteractive < ActiveRecord::Base
   # getters for inherited attributes
   def aspect_ratio_method
-    inherit_aspect_ratio_method ? library_interactive.aspect_ratio_method : custom_aspect_ratio_method
+    inherit_aspect_ratio_method && library_interactive ? library_interactive.aspect_ratio_method : custom_aspect_ratio_method
```
We talked about this and decided to leave it as is for now. We might want to tweak it later after we can play with the full experience more and see what happens when authors create interactives but don't give them a library interactive.
